pkg/covermerger/monorepo.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package covermerger
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/google/syzkaller/pkg/vcs"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type fileVersProvider interface {
+	GetFileVersions(c *Config, targetFilePath string, rbcs []RepoBranchCommit,
+	) (fileVersions, error)
+}
+
+type monoRepo struct {
+	branches map[RepoBranchCommit]struct{}
+	mu       sync.RWMutex
+	repo     vcs.Repo
+}
+
+type fileVersions map[RepoBranchCommit]string
+
+func (mr *monoRepo) GetFileVersions(c *Config, targetFilePath string, rbcs []RepoBranchCommit,
+) (fileVersions, error) {
+	mr.mu.RLock()
+	if !mr.allRepoBranchesPresent(rbcs) {
+		mr.mu.RUnlock()
+		if err := mr.cloneBranches(rbcs); err != nil {
+			return nil, fmt.Errorf("failed to clone repos: %w", err)
+		}
+		mr.mu.RLock()
+	}
+	defer mr.mu.RUnlock()
+	res := make(fileVersions)
+	for _, rbc := range rbcs {
+		fileBytes, err := mr.repo.Object(targetFilePath, rbc.Commit)
+		// It is ok if some file doesn't exist. It means we have repo FS diff.
+		if err != nil {
+			continue
+		}
+		res[rbc] = string(fileBytes)
+	}
+	return res, nil
+}
+
+func (mr *monoRepo) allRepoBranchesPresent(rbcs []RepoBranchCommit) bool {
```

Comment:
Nit: for the functions that are assumed to be only run under a lock, but don't take the lock themselves, it's common to append a `Locked` suffix to their names. It will make the intentions clearer for the reader.

Reply:
Let's discuss it offline.